src/pubtools/sign/clients/registry.py

```diff
@@ -0,0 +1,168 @@
+    "${REGISTRY_AUTH_FILE}",
+]
+
+
+@dataclasses.dataclass
+class AuthTokenWrapper:
+    """Carrier of the auth token for container registry."""
+
+    token: str
+
+
+class ContainerRegistryClient:
+    """Client for interacting with container registries.
+
+    Args:
+        username (Optional[str]): Username for authentication.
+        password (Optional[str]): Password for authentication.
+        auth_file (Optional[str]): Path to the auth file.
+    """
+
+    def __init__(
+        self,
+        username: Optional[str] = None,
+        password: Optional[str] = None,
+        auth_file: Optional[str] = None,
+        retries: int = 5,
+    ):
+        """Initialize.
+
+        Args:
+            username (Optional[str]): Username for authentication.
+            password (Optional[str]): Password for authentication.
+            auth_file (Optional[str]): Path to the auth file.
+        """
+        self.username = username
+        self.password = password
+        self.auth_file = auth_file
+        self._session = None
+        self.retries = retries
+
+    @property
+    def session(self):
+        """Get the session object."""
+        if not self._session:
+            self._session = requests.Session()
+            retries = Retry(
+                total=self.retries, backoff_factor=0.1, status_forcelist=[500, 502, 503, 504]
+            )
+            self._session.mount("http://", HTTPAdapter(max_retries=retries))
+        return self._session
+
+    def resolve_authentication(self, image_reference: str) -> Tuple[str, str]:
+        """Resolve authentication for the given image reference.
+
+        Configuration files for container authentication are searched in the
+        first place. If there's no match found in those, the username and
+        password provided to the client are used.
+
+        Args:
+            image_reference (str): Image reference to resolve authentication for.
+        Returns:
+            Tuple[str, str]: Username and password for authentication.
+        """
+        parsed = urlparse(image_reference)
+        if not parsed.scheme:
+            parsed = urlparse(f"docker://{image_reference}")
+        registry = parsed.netloc
+        existing_auth_files = []
+        auth_files = AUTH_FILES if not self.auth_file else [self.auth_file] + AUTH_FILES
+        for af in auth_files:
+            if os.path.exists(os.path.expandvars(af)):
+                existing_auth_files.append(os.path.expandvars(af))
+        for eaf in existing_auth_files:
+            parsed_conf = json.load(open(os.path.expandvars(eaf)))
+            parsed_auths = parsed_conf.get("auths")
+            if registry in parsed_auths:
+                auth = (
+                    base64.b64decode(parsed_auths.get(registry)["auth"].encode("utf-8"))
+                    .decode("utf-8")
+                    .split(":")
+                )
+                break
+        else:
+            auth = (self.username, self.password)
```

Comment:
It seems a bit odd to me that explicitly entering username and password will make them unused by default. Shouldn't explicitly specified arguments have precedence over config data? In what scenario would the current functionality be desirable?

Reply:
hmm, yeah, maybe it would be better to reverse it. I wanted username and password to be last priority for getting authentication values, but it's probably confusing as usually when you specify it it's used as override